src/nc_ssl.c

```diff
@@ -0,0 +1,301 @@
+        case SSL_ERROR_WANT_WRITE: log_error("error string: SSL_ERROR_WANT_WRITE"); break;
+        case SSL_ERROR_WANT_CONNECT: log_error("error string: SSL_ERROR_WANT_CONNECT"); break;
+        case SSL_ERROR_WANT_ACCEPT: log_error("error string: SSL_ERROR_WANT_ACCEPT"); break;
+        case SSL_ERROR_WANT_X509_LOOKUP: log_error("error string: SSL_ERROR_WANT_X509_LOOKUP"); break;
+        case SSL_ERROR_SYSCALL: log_error("error string: SSL_ERROR_SYSCALL"); break;
+        case SSL_ERROR_SSL: log_error("error string: SSL_ERROR_SSL"); break;
+    }
+
+    if (error_code == SSL_ERROR_SSL) {
+        log_ssl_error_stack();
+    }
+}
+
+static rstatus_t
+block_until_read_or_write(int socket_descriptor, int timeout_secs) {
+    fd_set fds;
+    struct timeval tv;
+
+    FD_ZERO(&fds);
+    FD_SET(socket_descriptor, &fds);
+    tv.tv_sec = timeout_secs;
+    tv.tv_usec = 0;
+    int select_result = select(socket_descriptor+1, &fds, &fds, NULL, &tv);
+
+    if (select_result <= 0) {
+        log_debug(LOG_VERB, "select() failed with code: %d", select_result);
+        return NC_ERROR;
+    }
+
+    return NC_OK;
+}
+
+static rstatus_t
+do_ssl_connect(SSL *ssl) {
+    int connect_status;
+    while ((connect_status = SSL_connect(ssl)) != 1) {
+        int code = SSL_get_error(ssl, connect_status);
+
+        if (code == SSL_ERROR_WANT_READ || code == SSL_ERROR_WANT_WRITE) {
+            // This means that the socket needs to do a read or write first.
+            // Since the socket is nonblocking, we can just wait until it is done, per the man page.
+            if (NC_OK != block_until_read_or_write(SSL_get_fd(ssl), 2)) {
+                log_error("Failing SSL_connect due to error waiting for read or write.");
+                return NC_ERROR;
+            }
+        }
+        else {
+            log_error("Failing SSL_connect due to unhandled error.");
+            log_ssl_error_code(code);
+            return NC_ERROR;
+        }
+    }
+
+    log_debug(LOG_VERB, "Successfuly completed SSL connection.");
+    return NC_OK;
+}
+
+rstatus_t
+nc_setup_ssl(struct conn *conn, struct string *host_cert_path, struct string *host_key_path, struct string *ca_file_path) {
+    char *cert_path = (char *)host_cert_path->data;
+    char *key_path = (char *)host_key_path->data;
+    char *ca_path = (char *)ca_file_path->data;
+
+    log_debug(LOG_VERB, "Using %s", SSLeay_version(SSLEAY_VERSION));
+    log_debug(LOG_VERB, "Connecting with cert: %s", cert_path);
+    log_debug(LOG_VERB, "Connecting with key: %s", key_path);
+    log_debug(LOG_VERB, "Connecting with ca file: %s", ca_path);
+
+    SSL_CTX *ctx;
+    SSL *ssl;
+
+    // FIXME: can these 3 functions be called multiple times?
+    SSL_library_init(); /* load encryption & hash algorithms for SSL */
+    ERR_load_crypto_strings();
+    SSL_load_error_strings(); /* load the error strings for good error reporting */
+
+    if ((ctx = SSL_CTX_new(SSLv23_method())) == NULL) {
```

Comment:
Question: does `SSLv23_method` result in a connection via SSL2/3, or does it also support TLS connections? We should only use TLS 1.1+, not SSL v2/v3.

I have a hunch this is just a badly named method from upstream, but just wanted to check.

Reply:
You're hunch is correct :)
From the [man page](https://www.openssl.org/docs/man1.0.2/ssl/SSL_CTX_new.html):
```
These are the general-purpose version-flexible SSL/TLS methods. The actual protocol version used will be 
negotiated to the highest version mutually supported by the client and the server. The supported protocols 
are SSLv2, SSLv3, TLSv1, TLSv1.1 and TLSv1.2.
```

We can pin it to a specific version of TLS, if that's preferable.